[new connector] Generic Database Connector (PostgreSQL, Oracle)

moxarth-elastic opened the pull request:
This PR includes:

Ping functionality with test cases.
Indexing functionality with configured connection string.
E2E test cases with the docker image.

## Checklists

 - [x] Add a module or a directory in [connectors/sources](https://github.com/elastic/connectors-python/tree/main/connectors/sources)
 - [x] Implement a class that implements all methods described in connectors.source.BaseDataSource
 - [x] Add a unit test in [connectors/sources/tests](https://github.com/elastic/connectors-python/tree/main/connectors/sources/tests) with +90% coverage
 - [x] Declare your connector in [config.yml](https://github.com/elastic/connectors-python/blob/main/config.yml) in the sources section
 - [x] Declare your dependencies in [requirements.txt](https://github.com/elastic/connectors-python/tree/main/requirements). Make sure you pin these dependencies
 - [x] Make sure you use an async lib for your source. If not possible, make sure you don't block the loop
 - [x] When possible, provide a docker image that runs the backend service, so we can test the connector. If you can't provide a docker image, provide the credentials needed to run against an online service.
 - [x] The test backend needs to return more than 10k documents due to 10k being a default size limit for Elasticsearch pagination. Having more than 10k documents returned from the test backend will help the testing connector more deeply.
 - [x] Added functional test to the nightly buildkite
 - [x] SSL Support for Oracle.

## Licenses

- SQLAlchemy: [MIT License](https://github.com/sqlalchemy/sqlalchemy/blob/main/LICENSE)
- asyncpg: [Apache-2.0 License](https://github.com/MagicStack/asyncpg/blob/master/LICENSE)
- cx_Oralce: [BSD License](https://github.com/oracle/python-cx_Oracle/blob/main/LICENSE.txt)

tarekziade commented:
Awesome will review it now -- would it be possible to try it out against Azure? (I can give you an access to an Azure SQL deployment if you need)

tarekziade commented on connectors/sources/mssql.py:
```diff
@@ -0,0 +1,85 @@
+        }
+        self.engine = None
+
+    async def ping(self):
```
It feels like the ping is copy-pasted around, and could be shared in the base class.

moxarth-elastic replied:
Here we need to pass `is_async` flag to `execute_query`. So what do you think can we bifurcate it via a type of engine in the common ping method in the base class?

moxarth-elastic replied:
Hey @tarekziade, should we move with your suggestion with the bifurcation of engine type at the common ping method?

moxarth-elastic replied:
We have accommodated ping in the generic file. Please check and let us know your thoughts.

tarekziade commented on connectors/sources/generic_database.py:
```diff
@@ -0,0 +1,444 @@
+        super().__init__(connector=connector)
+
+        # Connector configurations
+        self.retry_count = int(
```
with the latest changes, you can do:

```
self.retry_count = self.configuration['retry_count']
```

The base class is taking care of the default value and the type conversion now, as long as you set the right type in `get_default_configuration`

moxarth-elastic replied:
Done

tarekziade commented on connectors/sources/mssql.py:
```diff
@@ -0,0 +1,85 @@
+        self.connection_string = f"mssql+pymssql://{self.user}:{quote(self.password)}@{self.host}:{self.port}/{self.database}"
+        self.queries = {
+            "PING": "SELECT 1+1",
+            "ALL_TABLE": "SELECT table_name FROM information_schema.tables WHERE TABLE_SCHEMA = '{schema}'",
```
Can we move the queries as global consts to make it easier to read this code?

moxarth-elastic replied:
Done

tarekziade commented on connectors/sources/generic_database.py:
```diff
@@ -0,0 +1,444 @@
+            list: Column names and query response
+        """
+        query = self.queries[query_name].format(**query_kwargs)
+        size = int(self.configuration.get("fetch_size", DEFAULT_FETCH_SIZE))
```
same remark (it's now auto converted and defaulted)

tarekziade commented on connectors/sources/generic_database.py:
```diff
@@ -0,0 +1,444 @@
+            )
+            raise
+
+    def serialize(self, doc):
```
this is being copy-pasted in many sources at this point.

Let's move it to `BaseDataSource` to avoid duplication (and remove it from mongo, MySQL)

tarekziade commented on connectors/sources/generic_database.py:
```diff
@@ -0,0 +1,444 @@
+                    table=table,
+                )
+            )
+            if row_count:
```
`row_count` is not a boolean and a rather complex structure.. (`[[]]`) can we change this for an explicit test?

moxarth-elastic replied:
Done

tarekziade commented on connectors/sources/generic_database.py:
```diff
@@ -0,0 +1,444 @@
+                schema=schema,
+            )
+        )
+        if list_of_tables:
```
explicit test, `list_of_tables` is not a bool

tarekziade commented on connectors/sources/oracle.py:
```diff
@@ -0,0 +1,58 @@
+        self.queries = {
+            "PING": "SELECT 1+1 FROM DUAL",
+            "ALL_TABLE": "SELECT TABLE_NAME FROM all_tables where OWNER = '{user}'",
+            "TABLE_PRIMARY_KEY": "SELECT cols.column_name FROM all_constraints cons, all_cons_columns cols WHERE cols.table_name = '{table}' AND cons.constraint_type = 'P' AND cons.constraint_name = cols.constraint_name AND cons.owner = cols.owner ORDER BY cols.table_name, cols.position",
```
same remark (globals)

tarekziade commented on connectors/sources/postgresql.py:
```diff
@@ -0,0 +1,81 @@
+        self.queries = {
+            "PING": "SELECT 1+1",
+            "ALL_TABLE": "SELECT table_name FROM information_schema.tables WHERE table_catalog = '{database}' and table_schema = '{schema}'",
+            "TABLE_PRIMARY_KEY": "SELECT c.column_name FROM information_schema.table_constraints tc JOIN information_schema.constraint_column_usage AS ccu USING (constraint_schema, constraint_name) JOIN information_schema.columns AS c ON c.table_schema = tc.constraint_schema AND tc.table_name = c.table_name AND ccu.column_name = c.column_name WHERE constraint_type = 'PRIMARY KEY' and tc.table_nam ...
```
const

tarekziade commented on connectors/sources/postgresql.py:
```diff
@@ -0,0 +1,81 @@
+        Returns:
+            dictionary: Connection arguments
+        """
+        self.ssl_ca = self.ssl_ca.replace(" ", "\n")
```
Why do we do some specific extraction here? Shouldn't that be done in the generic connector? it looks like you are removing the `--BEGIN` etc? should be a function with an explicit name, like `parse_pem` that returns the key etc

moxarth-elastic replied:
> Why do we do some specific extraction here?

This method is specific to PostgreSQL dialect only, so we should not move it to generic connector file.

> Shouldn't that be done in the generic connector? it looks like you are removing the --BEGIN etc?

Also, we're not removing `---BEGIN--` and `---END---` from the string, we're making it like `pem` format to parse it into `cadata` - https://github.com/elastic/connectors-python/blob/c65bd1ea4b538cbc395f47d654aa2813cf2ec826/connectors/sources/postgresql.py#L65

> should be a function with an explicit name, like parse_pem that returns the key etc

The function takes care of creating a connection argument for engine that's why we named this way.

tarekziade replied:
can you create a separate 

```
pem_format = self.get_pem_format(self.ssl_ca)
```

with a unit test that passes a SSL cert and returns the pem format.
This will clarify what we are doing here.

thx

moxarth-elastic replied:
Done

moxarth-elastic commented:
> Awesome will review it now -- would it be possible to try it out against Azure? (I can give you an access to an Azure SQL deployment if you need)

Sure @tarekziade, we can try it out for Azure SQL. You can slack me the creds for the same.

tarekziade commented on requirements/aarch64.txt:
```diff
@@ -5,3 +5,7 @@ aiomysql==0.0.22
 aioboto3==9.0.0
 motor==2.5.1
 smbprotocol==1.9.0
+SQLAlchemy==1.4.44
```
since it's the exact same list than x86_64, let's move them to the framework.txt one

we also need for each new lib to check the license, if you can list them in the issue, so we check that we don't introduce a lib that is not in the list of allowed licenses for Elastic thx

moxarth-elastic replied:
Ok move all libs to framework.txt

tarekziade reviewed (commented):
just one nit or two and we can land
thanks for all the small fixes!
the most important one is licensing

tarekziade commented on connectors/sources/mssql.py:
```diff
@@ -0,0 +1,87 @@
+
+
+class MSSQLDataSource(GenericBaseDataSource):
+    """Class to fetch documents from Microsoft SQL Server"""
```
this docstring is used in the CLI, let's tweak the name, see https://github.com/elastic/connectors-python/pull/348

moxarth-elastic replied:
Done

tarekziade commented on connectors/sources/oracle.py:
```diff
@@ -0,0 +1,60 @@
+
+
+class OracleDataSource(GenericBaseDataSource):
+    """Class to fetch documents from Oracle Server"""
```
this docstring is used in the CLI, let's tweak the name, see https://github.com/elastic/connectors-python/pull/348

tarekziade commented on connectors/sources/postgresql.py:
```diff
@@ -0,0 +1,83 @@
+
+
+class PostgreSQLDataSource(GenericBaseDataSource):
+    """Class to fetch documents from Postgresql Server"""
```
this docstring is used in the CLI, let's tweak the name, see https://github.com/elastic/connectors-python/pull/348

tarekziade reviewed (changes requested):
I am sorry we have one more extra step before we can merge, sorry this did not come on my radar before:

- pymssql: [LGPL-2.1](https://github.com/pymssql/pymssql/blob/master/LICENSE)
- psycopg2: [LGPL](https://github.com/psycopg/psycopg2/blob/master/LICENSE)

These are LGPL and we need a specific process to allow heir usage. 

cc @gtback for gudiance

tarekziade commented:
Also relevant https://github.com/pymssql/pymssql/issues/668

@moxarth-elastic could you try

- https://github.com/mkleehammer/pyodbc (MIT) 
- https://pypi.org/project/asyncpg/ (APLv2)

moxarth-elastic commented:
Hey @tarekziade,

> https://pypi.org/project/asyncpg/ 

we have removed the package psycopg2 and updated our e2e implementation with the new suggested package.

> Also relevant pymssql/pymssql#668

as per their latest comment https://github.com/pymssql/pymssql/issues/668#issuecomment-1086681292 the decision has been reversed by the author and it will be continued. SQLAlchemy also allows pymssql [Reference](https://github.com/sqlalchemy/sqlalchemy/issues/8337)

Also, we did check PyODBC at the early stage of development [doc](https://docs.google.com/document/d/1Hf8ARyw9kXBqQ25FKQK36AMKEMVfRm5j7ssIPp8zvfc/edit?usp=drivesdk). However, we decided to not go through it for the following reasons:

- The prerequisite to use the connector will be an ODBC driver. The user would have to ensure the required drivers are installed and they pass the path to the driver as a connector configuration. Also, the installation of the driver may vary for the different OS as well as different versions of it.

However, if we decide to still move ahead with pyodbc, then we'll have to make the necessary changes and also retest the connector for functional, regression, as well as performance testing against the Mssql dialect. We anticipate this to take about an additional week approximately to develop, test, and update the PR. Please let us know your thoughts.

CC: @danajuratoni, @nirav-elastic

moxarth-elastic commented:
> Using LGPL is a blocker from a licensing Pov unfortunately. This is what I suggest
> 
> * let's remove mssql from this patch so we can land it, it has been around for a while
> * let's add mssql as a follow-up patch, and we can dig into the problem

Sure let me remove the mssql dialect so we can merge this PR.
Also, we are string implementation with pyodbc for mssql dialect and we will the follow up PR in the upcoming week.

tarekziade commented:
Thanks a lot!

one nit before we can merge: @artem-shelkovnikov added a linter that requires tests to pass under a second, we need to make sure those 2 tests do this:


```
FAILED connectors/sources/tests/test_oracle.py::test_ping_negative
FAILED connectors/sources/tests/test_postgresql.py::test_ping_negative
```

They take >6 seconds. You can look at how we sped up the azure and GCS ones for the pings -- it's just a matter of tweaking the config. I would do it myself but I can't access your branch

moxarth-elastic commented:
> Thanks a lot!
> 
> one nit before we can merge: @artem-shelkovnikov added a linter that requires tests to pass under a second, we need to make sure those 2 tests do this:
> 
> ```
> FAILED connectors/sources/tests/test_oracle.py::test_ping_negative
> FAILED connectors/sources/tests/test_postgresql.py::test_ping_negative
> ```
> 
> They take >6 seconds. You can look at how we sped up the azure and GCS ones for the pings -- it's just a matter of tweaking the config. I would do it myself but I can't access your branch

Hey @tarekziade,

I have updated the test cases. please take a look.

gtback commented:
> cc @gtback for gudiance

Sorry I missed the ping here. If you still need anything from me, let me know.
